### PR TITLE
[WIP] patch to add support for array by table subquery

### PR DESCRIPTION
**Pull request by JamesRTaylor**
Fixes https://github.com/prestosql/presto/issues/12

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java [hunk 1]
```diff
@@ -297,6 +318,30 @@ private PlanBuilder appendExistSubqueryApplyNode(PlanBuilder subPlan, ExistsPred
                 correlationAllowed);
     }
 
+    private PlanBuilder appendArraySubqueryApplyNode(PlanBuilder subPlan, ArrayConstructor arraySubquery, boolean correlationAllowed)
```
Comment (JamesRTaylor) [inline presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java]: Not sure what I need to do here. Can I wrap the projection with an ARRAY_AGG here, or do I need to do the desugar later during PlanOptimizers?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DesugarArrayConstructorSubquery.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    private static ExpressionRewriter createRewrite(Metadata metadata)
+    {
+        return (expression, context) -> ExpressionTreeRewriter.rewriteWith(new io.prestosql.sql.tree.ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteArrayConstructor(ArrayConstructor node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                if (node.isSubquery()) {
+                    SubqueryExpression subquery = node.getSubquery();
+                    // TODO: wrap single projected node with ARRAY_AGG
```
Comment (JamesRTaylor) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DesugarArrayConstructorSubquery.java]: Is this the right place to rewrite the subquery to wrap the projected node with an ARRAY_AGG?

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/ApplyNode.java [hunk 3]
```diff
@@ -101,7 +102,8 @@ private static boolean isSupportedSubqueryExpression(Expression expression)
     {
         return expression instanceof InPredicate ||
                 expression instanceof ExistsPredicate ||
-                expression instanceof QuantifiedComparisonExpression;
+                expression instanceof QuantifiedComparisonExpression ||
+                expression instanceof ArrayConstructor;
```
Comment (JamesRTaylor) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/ApplyNode.java]: Unclear to me how I'll use this. Will I use this to pattern match my rules?

File: presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java [hunk 4]
```diff
@@ -87,6 +87,20 @@ public void testUnsupportedSubqueriesWithCoercions()
                 UNSUPPORTED_CORRELATED_SUBQUERY_ERROR_MSG);
     }
 
+    @Test
+    public void testSimpleArraySubqueries()
```
Comment (JamesRTaylor) [inline presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java]: Super simple test - will of course add many more once I get the basic flow working.

File: presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4 [hunk 5]
```diff
@@ -488,7 +489,7 @@ number
 
 nonReserved
     // IMPORTANT: this rule must only contain tokens. Nested rules are not supported. See SqlParser.exitNonReserved
-    : ADD | ADMIN | ALL | ANALYZE | ANY | ARRAY | ASC | AT
+    : ADD | ADMIN | ALL | ANALYZE | ANY | ASC | AT
```
Comment (JamesRTaylor) [inline presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4]: Without this change, the ARRAY(...) was being interpreted by the parser as a function call.

File: presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4 [hunk 6]
```diff
@@ -308,6 +308,7 @@ primaryExpression
     | identifier '->' expression                                                          #lambda
     | '(' (identifier (',' identifier)*)? ')' '->' expression                             #lambda
     | '(' query ')'                                                                       #subqueryExpression
+    | ARRAY '(' query ')'                                                                 #arrayByTableSubquery
```
Comment (findepi) [inline presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4]: Is it equivalent to `SELECT array_agg(x) FROM query`?

Comment (martint) [reply presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4]: Yes, as a scalar subquery.

File: presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java (hunk 1 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java]: The way we deal with these (IN, EXISTS, etc) today is not great, but let's model it like those: we add an `Apply` node whose "assignment expression" is rewritten to be "x = ARRAY (y)", where y is the symbol name associated with the output of the subquery. So, what you have here is good.

We'll also need a rule similar to `TransformUncorrelatedInPredicateSubqueryToSemiJoin` that converts an Apply containing an Array(..) expression into a join + scalar subquery with with `array_agg`.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DesugarArrayConstructorSubquery.java (hunk 2 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DesugarArrayConstructorSubquery.java]: Take a look at `TransformUncorrelatedInPredicateSubqueryToSemiJoin`

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/ApplyNode.java (hunk 3 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/ApplyNode.java]: This is mostly for sanity checking. It's to make sure we don't inadvertently create Apply nodes with expressions that are not supported.

**Comment by bitsondatadev** [comment]
:wave: @JamesRTaylor - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**